lang/funcs/collection.go

```diff
@@ -1052,50 +1062,80 @@ var SliceFunc = function.New(&function.Spec{
 		},
 	},
 	Type: func(args []cty.Value) (cty.Type, error) {
-		return args[0].Type(), nil
-	},
-	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
-		inputList := args[0]
-		if !inputList.IsWhollyKnown() {
-			return cty.UnknownVal(retType), nil
-		}
-		var startIndex, endIndex int
+		arg := args[0]
+		argTy := arg.Type()
 
-		if err = gocty.FromCtyValue(args[1], &startIndex); err != nil {
-			return cty.NilVal, fmt.Errorf("invalid start index: %s", err)
+		if !args[0].IsKnown() {
+			return argTy, nil
 		}
-		if err = gocty.FromCtyValue(args[2], &endIndex); err != nil {
-			return cty.NilVal, fmt.Errorf("invalid start index: %s", err)
+
+		if !argTy.IsListType() && !argTy.IsTupleType() {
+			return cty.NilType, fmt.Errorf("can only flatten lists, and tuples")
```

Comment:
I think it could be helpful to also have a special case here for sets so we can give a more actionable error:

```go
if argTy.IsSetType() {
    return cty.NilType, fmt.Errorf("cannot slice a set, because its elements do not have indices; use the tolist function to force conversion to list if the ordering of the result is not important")
}
```

(The comma in this existing message also feels a little unnatural.)